common/zeroconf/zeroconf.cpp

```diff
@@ -0,0 +1,130 @@
+
+namespace Aseba
+{
+	using namespace Dashel;
+
+	//! Register and announce a target described by a name and a port
+	void Zeroconf::registerPort(const string name, const unsigned port, const string txt)
+	{
+        ZeroconfService z;
+        uint16_t len = txt.size();
+        const char* record = txt.c_str();
+        DNSServiceErrorType err = DNSServiceRegister(&z.serviceref,
+                                                     kDNSServiceFlagsDefault,
+                                                     0, // default all interfaces
+                                                     name.c_str(),
+                                                     "_aseba._tcp",
+                                                     NULL, // use default domain
+                                                     NULL, // use this host name
+                                                     htons(port),
+                                                     len, // TXT length
+                                                     record, // TXT record
+                                                     cb_Register,
+                                                     &z);
+        if (err != kDNSServiceErr_NoError)
+            throw runtime_error(FormatableString("Zeroconf DNSServiceRegister: error %0").arg(err));
+        else
+        {
+            DNSServiceErrorType err = DNSServiceProcessResult(z.serviceref); // block until daemon replies
+            if (err != kDNSServiceErr_NoError)
+                throw runtime_error(FormatableString("Zeroconf DNSServiceProcessResult: error %0").arg(err));
+            else
+            {
+                services[z.name] = z; // copy
+                z.serviceref = 0; // because now owned by services
+            }
+        }
+	}
+
+    //! Register and announce a target described by an existing Dashel stream
+    void Zeroconf::registerStream(const std::string name, const Dashel::Stream* stream, const string txt)
+    {
+        string service_name = name;
+        unsigned port = atoi(stream->getTargetParameter("port").c_str());
+        registerPort(name,port,txt);
+    }
+
+    //! Asynchronously update the set of known targets
+    void Zeroconf::browseForTargets()
+    {
+    }
+
+    //! callback to update ZeroconfService record with results of registration
+    void DNSSD_API Zeroconf::cb_Register(DNSServiceRef sdRef,
+                               DNSServiceFlags flags,
+                               DNSServiceErrorType errorCode,
+                               const char *name,
+                               const char *regtype,
+                               const char *domain,
+                               void *context)
+    {
+        if (errorCode != kDNSServiceErr_NoError)
+            throw runtime_error(FormatableString("Zeroconf DNSServiceRegisterReply: error %0").arg(errorCode));
+        else
+        {
+            ZeroconfService *zref = (ZeroconfService *)context;
```

Comment:
`Aseba::polymorphic_downcast<T>(base)` (defined in `common/utils/utils.h`) should be used instead of C-style cast.